---
title: "Proteomics Analysis Pipeline Results"
output:
  html_document:
    toc: TRUE
    toc_float: TRUE
    theme: spacelab
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r, warning=FALSE, message=FALSE}
library(dplyr)
library(drake)
library(enrichplot)
library(DT)
library(kableExtra)
library(knitr)
```
  
```{r load targets}
loadd(data)
loadd(wilcoxonResults)
loadd(foldChangeResults)
loadd(plotResults)
loadd(randomForestResults)
loadd(proteinsRanked)
loadd(proteinSetUnique)
loadd(proteinID)
loadd(GOAnalysisResults)
loadd(DOAnalysisResults)
loadd(pathwayAnalysisResults)
loadd(GOAnalysisPlot)
# loadd(DOAnalysisPlot)
# loadd(pathwayAnalysisPlot)
```

## Data

```{r data info display}
names <- c("Number of features", "Number of samples", "Case class", "Control class",
  "Number of samples in case class", "Number of samples in control class")

values <- as.character(c(dim(data)[2] - 2, dim(data)[1], levels(data$Class)[1], 
  levels(data$Class)[2],
  nrow(data[which(data[, 2] == levels(data[, 2])[1]), ]),
  nrow(data[which(data[, 2] == levels(data[, 2])[2]), ])))

info <- cbind(names, values)

DT::datatable(info, options = list(dom = 't', ordering = FALSE), colnames = rep("", ncol(info)), )
```

## Univariate analysis {.tabset}

### Wilcoxon Rank-sum Test
```{r wilcoxon results NA check}

#check for NA values in data frame
wilcoxonCounterNA <- sum(is.na(wilcoxonResults[, 4]))
  
#number of proteins for which p-value was calculated
N <- nrow(wilcoxonResults) - wilcoxonCounterNA

#calculate number of significant proteins
wilcoxonResultsSignificant <- wilcoxonResults %>% 
  filter(Pvalue.adj < 0.05) %>% 
  arrange(Pvalue.adj)

M <- nrow(wilcoxonResultsSignificant)
```

The wilcoxon test failed for `r wilcoxonCounterNA` proteins. 

P-value calculated for **`r N`** proteins. Out of those **`r M`** proteins showed a significant p-value < 0.05. The complete list of significant proteins can be found below.

```{r wilcoxon results table DT}
DT::datatable(wilcoxonResultsSignificant, colnames = c("Description", "Gene", "P-value", 
  "Adjusted P-value")) %>% 
  formatRound(c("Pvalue", "Pvalue.adj"), 6)
```

### Fold Change 
```{r fold change results NA check}
#check for NA values in data frame
foldChangeCounterNA <- sum(is.na(foldChangeResults[, 2]))
```

The fold change calculation failed for `r foldChangeCounterNA` proteins. 

The complete list of the calculated fold change between affected and control samples can be found below.

```{r fold change results table}
DT::datatable(foldChangeResults, colnames = c("Description", "Gene", "Fold change")) %>% 
  formatRound("foldChange", 6)
```

### Volcano Plot
```{r volcano plot}
plotResults
```

### Info

**Wilcoxon rank-sum test**

The Wilcoxon rank-sum test is a non-parametric statistical test that can be used to find significant differences between two unpaired groups that are not normally distributed. The observations of the two groups are combined and ordered by their value. A rank is assigned to each from low to high and the mean rank of both groups is then compared. A bigger difference between the mean ranks of the two groups, will produce a smaller and therefore more significant P-value^1^. The P-value is adjusted for multiple testing using the FDR method. 

In this analysis the mean ranks of the two classes `r levels(data[, 2])[1]` and `r levels(data[, 2])[2]` are compared.

**Fold change**

The calculation of the fold change for every feature, i.e. protein, allows to compare the difference in gene expression of a case group compared to a control group used as a baseline. It allows the identification of differentially-expressed genes between the two conditions^2^. As the input data is already expected to be log2-transformed, the fold change is simply the difference between the means of the control group, `r levels(data[, 2])[2]`, and the case group, `r levels(data[, 2])[1]`. 

$$FC_i = mean(X_i) - mean(Y_i)$$

A negative value for the fold change of a gene signals a downregulation of the expression, a positive fold change value signals upregulation.

**Volcano plot**

The volcano plot is a visualization method to display statistical significance of a difference versus the magnitude of change. This is achieved by plotting the log2 fold change on the x-axis and the negative log10 of the adjusted P-value on the y-axis for every protein in the data set^3^. The dotted lines mark a significant adjusted P-value (horizontally) or a gene expression in the case group increased or decreased by 0.2 (vertically). Features that meet both those conditions are highlighted in colour.

## Random Forest Classification {.tabset}

### Model Accuracy
```{r random forest accuracy}
testError <- vector()

for (i in 1:length(randomForestResults)) {
    #subset one random forest object
    fit.RF <- randomForestResults[[i]]
    #extract feature importance ranking and append to results vector
    testError <- append(testError, fit.RF$test$err.rate[fit.RF$ntree])
}

accuracy <- sapply(testError, FUN = function(x) (1 - x)*100)
```

The classification models produced an average accuracy of **`r round(mean(accuracy), 2)`** with a standard deviation of **`r round(sd(accuracy), 2)`**. The distribution of the individual model accuracies is shown in the following boxplot.

```{r random forest accuracy plot}
ggplot(as.data.frame(accuracy), aes(x="", y = accuracy)) + 
  geom_boxplot(colour = "black", fill = "dodgerblue4") +
  labs(x = "", y = "accuracy", title = "Classification accuracy of Random Forest models")
```

### Protein Ranking
```{r proteins ranked table}
DT::datatable(proteinsRanked, rownames = FALSE) %>% 
  formatRound("Importance", 6)
```

### Protein Set Selection

`r nrow(proteinSetUnique)` were selected. Out of those **`r nrow(proteinID)`** could be mapped to their unique Entrez ID and will be used for the enrichment analyses.

```{r proteins selected importance table}
DT::datatable(proteinID, rownames = FALSE, escape = FALSE)
```


### Highest Ranking Proteins
```{r proteins selected importance plot}
#subset for top 10 features 
topTenGenes <- proteinsRanked[1:10, 2:3]

#reset order of levels in factor
topTenGenes$Gene <- factor(topTenGenes$Gene,
                        levels=topTenGenes$Gene)

ggplot(topTenGenes, aes(x = topTenGenes[, 1], y = topTenGenes[, 2])) +
  geom_bar(aes(), stat="identity", colour = "black", fill = "dodgerblue4") +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
    labs(x = "Genes", y = "Importance", title = 
      "Genes with highest average importance across all Random Forest models")
```

### Info 

**Random Forest**

Random forest is a supervised machine learning method that can be used for classification problems. The aim of the model in this pipeline is to correctly predict for an observation which class it belongs to based on the protein measurements of that observation.

The method works by segmenting the feature space into simpler regions that allow the assignment of an observation to a class with increasing confidence. The rules for splitting the feature space can be summarized in so-called decision trees. To build such a decision tree, the model searches for the feature variable that is best suitable to split the observations in a way that separates the two classes in the data set the best. After a first split, two non-overlapping subregions of the feature space exist and the model searches for the next feature to further separate the regions according to their classes. This step is repeated again and again to grow the tree and to segment the feature space into regions of higher class purity^4^.

Random forest use two strategies to improve the accuracy of this tree-based classifiaction. As the name suggests, the class prediction for an observation is based on many decision trees instead of just one. The consensus of all trees is the class chosen by the model. Moreover, before each tree for the model is built a random sampling of the features is performed. As a result, every tree can only use a small subset of the available features which ensures that the trees in a random forest are different from and uncorrelated to each other^4^. 

In the pipeline 20 random forest models were built each containing 6000 trees. 

For more information on the algorithm used, see the randomForest R package documentation: https://www.rdocumentation.org/packages/randomForest/versions/4.6-14/topics/randomForest.

**Protein Ranking**

Random forest allows to rank all variables used based on their importance for the classification model. The ranking is based on the mean decrease in Gini index. The Gini index is considered a measure of node purity and is defined as

$$G = \sum_{k=1}^K p_{mk}(1-p_{mk})$$

with $p_{mk}$ representing the proportion of observations in the m^th^ region that are from the k^th^ class^4^.

The decrease in Gini index measures the decrease in impurity that a feature facilitates as it is used in a split of the feature space and therefore measures the impact of this variable on the classification. The mean decrease in Gini index is the average of the impurity reductions over all trees in which the variable was used for classification. The higher the decrease in Gini index, the higher the importance of that variable in the model^5^.

The ranking of the proteins is based on all 20 random forest models fitted by averaging the mean decrease in Gini index of all forests.

**Protein Set Selection**

The selection of proteins that are analysed further depends on their mean decrease in Gini index relative to the feature with the highest importance. Every variable with an importance measure of at least one third of the highest ranking protein is included. If less than 50 proteins were selected by this measurement, the list is expanded using the results of the Wilcoxon test. Duplicates of the same protein are removed at this stage.


## Protein Set Analysis {.tabset}

### Gene Ontology Term Analysis {.tabset}

#### Table
```{r GO term analysis table}
DT::datatable(as.data.frame(GOAnalysisResults), rownames = FALSE, 
  colnames = c("ID", "Description",   "Gene Ratio", "BG Ratio", "P-value", "Adjusted P-value", 
  "Q-value", "Gene ID", "Count")) %>% 
  formatRound(c("pvalue", "p.adjust", "qvalue"), 6)
```

#### Bar Plot
```{r GO term analysis plot}
barplot(GOAnalysisResults)
```

#### Enrichment Map
```{r GO term analysis map}
emapplot(GOAnalysisResults)
```


### Disease Ontology Term Analysis {.tabset}

#### Table
```{r DO term analysis table}
DT::datatable(data.frame(DOAnalysisResults), rownames = FALSE, 
  colnames = c("ID", "Description", "Gene Ratio", "BG Ratio", "P-value", "Adjusted P-value", 
  "Q-value", "Gene ID", "Count")) %>% 
  formatRound(c("pvalue", "p.adjust", "qvalue"), 6)
```

#### Bar Plot
```{r DO term analysis plot}
barplot(DOAnalysisResults)
```

#### Enrichment Map
```{r DO term analysis map}
emapplot(DOAnalysisResults)
```

### Pathway Enrichment Analysis {.tabset}

#### Table
```{r pathway enrichment analysis table}
DT::datatable(data.frame(pathwayAnalysisResults), rownames = FALSE, 
  colnames = c("ID", "Description", "Gene Ratio", "BG Ratio", "P-value", "Adjusted P-value", 
  "Q-value", "Gene ID", "Count")) %>% 
  formatRound(c("pvalue", "p.adjust", "qvalue"), 6)
```

#### Bar Plot
```{r pathway enrichment analysis plot}
barplot(pathwayAnalysisResults)
```

#### Enrichment Map
```{r pathway enrichment analysis map}
emapplot(pathwayAnalysisResults)
```

### Info

While the univariate analyses and machine learning model can deliver a list of proteins of interest, they are lacking any context on their functions and interactions. We therefore use enrichment analyses to infer actual knowledge from the set of proteins. The general workflow of such an approach is the systematical mapping of the proteins to their associated annotations, followed by a comparison of the distribution of terms within the protein set with a predefined background^6^.

**Gene Ontology Term Enrichment Analysis**

GO is a defined vocabulary that allows annotation of all genes and proteins in a commonly agreed on language regarding their role and location^7^. GO terms are categorized into three categories: biological processes, molecular functions, and cellular components. Each category contains a hierarchical structure with the terms increasing in specificity. The GO term enrichment analysis maps a list of genes to their associated GO terms and performs an enrichment test for those terms. The enrichGo function of the R package clusterProfiler was used to perform this method^8^.

**Disease Ontology Term Enrichment Analysis**

Disease ontology (DO) is the analogous vocabulary for human diseases providing consistent and structured terms to describe and categorize them^9^. This ontology can be of special importance to give the protein list clinical relevance. The R package DOSE was used by implementing its the enrichDO function^10^.

**Pathway Enrichment Analysis**

Pathway enrichment analysis is used to determine if members of a pathway are overrepresented in the protein set. The ReactomePA package was used for this enrichment analysis and uses the Reactome database for annotation^11^. The Reactome database is manually curated and peer-reviewed pathway database that can facilitate the identification of high-order biological pathways in the data^12^.


## References

1 McDonald JH. 2014. Handbook of Biological Statistics (3rd ed.), pp.157-164. Sparky House
Publishing, Baltimore, Maryland.

2 Witten DM, Tibshirani R. 2007. A comparison of fold-change and the t-statistic
for microarray data analysis. Stanford University. 1-13.

3 Doyle M. 2019 Visualization of RNA-Seq results with Volcano Plot (Galaxy Training Materials). https://galaxyproject.github.io/training-material/topics/transcriptomics/tutorials/rna-seq-viz-with-volcanoplot/tutorial.html. Accessed: 17th April 2020.

4 James G, Witten D, Hastie T, Tibshirani R. 2013. An Introduction to Statistical Learning: with
Applications in R (8^th^ ed.). Springer Science+Business Media, New York.

5 Archer KJ, Kimes RV. 2008. Empirical characterization of random forest variable importance measures. Computational Statistics & Data Analysis, 52(4), 2249–2260. doi: 10.1016/j.csda.2007.08.015.

6 Tipney H, Hunter L. 2010. An introduction to effective use of enrichment analysis software. Hum Genomics 4, 202. doi: 10.1186/1479-7364-4-3-202.

7 Ashburner M, Ball CA, Blake JA et al. Gene ontology: tool for the unification of biology. Nat Genet. May 2000;25(1):25-9. doi: 10.1038/75556.

8 Yu G, Wang L, Han Y and He Q. 2012. clusterProfiler: an R package for comparing biological themes among gene clusters. OMICS: A Journal of Integrative Biology. 16(5):284-287. doi: 10.1089/omi.2011.0118.

9 Schriml, LM, Mitraka, E, Munro, J et al. 2018. Human Disease Ontology 2018 update: classification, content and workflow expansion. Nucleic acids research, 47(D1):D955–D962. doi: 10.1093/nar/gky1032.

10 Yu G, Wang L, Yan G, He Q. 2015. DOSE: an R/Bioconductor package for Disease Ontology Semantic and Enrichment analysis. Bioinformatics, 31(4), 608-609. doi: 10.1093/bioinformatics/btu684.

11 G Yu, QY He. 2016. ReactomePA: an R/Bioconductor package for reactome pathway analysis and visualization. Molecular BioSystems, 12(2):477-479. doi: 10.1039/C5MB00663E

12 Jassal B, Matthews L, Viteri G et al. 2020. The reactome pathway knowledgebase. Nucleic Acids Res. 48(D1):D498-D503. doi: 10.1093/nar/gkz1031.

